A33-Fault-Injection.md

```diff
@@ -0,0 +1,253 @@
+
+The way to implement the tracking of RPC lifespan also depends on the language. For example, Core & Java may use the deallocation method of the RPC object or enforcer. Go’s Finalizer might not work well enough, so the enforcer (interceptor) needs to update the active fault counter when RPC is finished.
+
+
+#### Header matching
+
+This doc proposes to support as many metadata as we can (see [detailed definitions](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/fault_filter.html?highlight=fault%20injection#controlling-fault-injection-via-http-headers)):
+
+```bash
+# Aborts HTTP requests with status code 503
+x-envoy-fault-abort-request=503
+
+# Aborts gRPC requests with INVALID_ARGUMENT
+X-envoy-fault-abort-grpc-request=3
+
+# The percentage of requests to be failed. This header only sets numerator. The denominator was set in the HTTPFault filter, which by default is 100.
+# 5% of requests should fail.
+X-envoy-fault-abort-request-percentage=5
+
+# Delays request for 53 milliseconds.
+X-envoy-fault-delay-request=53
+
+# The percentage of requests to be delayed. This header only sets numerator. The denominator was set in the HTTPFault filter, which by default is 100.
+# 20% of requests should delay
+X-envoy-fault-delay-request-percentage=20
+
+# Not supported
+X-envoy-fault-throughput-response=''
+X-envoy-fault-throughput-response-percentage=''
+```
+
+**Fault inject config in HTTP filter and headers won’t be conflicted**, because the header only being evaluated if the delay or abort is set to `HeaderDelay` or `HeaderAbort` instead of concrete values.
+
+
+### Status Code Mapping
+
+`grpc_status` is only available in xDS v3 API. In xDS v2 API, no matter TD or Envoy, they all only allow `http_status`. Mapping between HTTP status code and gRPC status code is well defined (see [doc](https://github.com/grpc/grpc/blob/master/doc/http-grpc-status-mapping.md)), but that means users are limited to inject only 7 out of 16 gRPC status codes (including `OK`). The underlying schema for abort looks like [FaultAbort](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/fault/v3/fault.proto#extensions-filters-http-fault-v3-faultabort). **This doc proposes to update TD configuration to allow more expressive status code injection, once xDS v3 API is rolled out.**
+
+We still accept fault injection settings with only HTTP status code. **The ConfigSelector should translate the HTTP status code into gRPC status code** before sending the config to the fault injection enforcer.
+
+
+### No multiple fault injection filter
```

Comment:
This section can be removed.  As per the HTTP filter design, we *will* support multiple instances of the same filter.